src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -0,0 +1,134 @@
+open Core_kernel
+open Async_kernel
+
+type 'a t =
+  { mvar: 'a option Mvar.Read_only.t
+  ; mutable cache: 'a option
+  ; mvar_pipe_handle: 'a option Pipe.Reader.t
+  ; mutable id: int
+  ; pipes:
+      ( 'a option Strict_pipe.Reader.t
+      * ( 'a option
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t )
+      Int.Table.t }
+
+let create mvar =
+  let mvar_pipe_handle = Mvar.pipe_when_ready mvar in
+  let t =
+    { mvar
+    ; cache= Mvar.peek mvar |> Option.join
+    ; mvar_pipe_handle
+    ; id= 0
+    ; pipes= Int.Table.create () }
+  in
+  don't_wait_for
+    (Pipe.iter_without_pushback t.mvar_pipe_handle ~f:(fun a ->
+         t.cache <- a ;
+         Int.Table.iteri t.pipes ~f:(fun ~key ~data:(_, w) ->
+             Core.printf !"Pipe%d %{sexp: int option}\n%!" key (Obj.magic a) ;
+             Strict_pipe.Writer.write w a ) )) ;
+  t
+
+let peek {cache; _} = cache
+
+let close t =
+  Core.printf !"Close all\n%!" ;
+  Pipe.close_read t.mvar_pipe_handle ;
+  Int.Table.iter t.pipes ~f:(fun (_, w) -> Strict_pipe.Writer.close w) ;
+  Int.Table.clear t.pipes
+
+let fresh_id t =
+  t.id <- t.id + 1 ;
+  t.id
+
+let prepare_pipe ~close t =
+  let r, w =
+    Strict_pipe.create
+      (Strict_pipe.Buffered (`Capacity 3, `Overflow Strict_pipe.Drop_head))
+  in
+  let id = fresh_id t in
+  Int.Table.add_exn t.pipes ~key:id ~data:(r, w) ;
+  don't_wait_for
+    (let%map () = close in
+     Core.printf !"Closing individual %d\n%!" id ;
+     Int.Table.remove t.pipes id ;
+     Strict_pipe.Writer.close w) ;
+  r
+
+let fold ~close t =
+  let r = prepare_pipe ~close t in
+  Strict_pipe.Reader.fold r
+
+let fold_without_pushback ~close t =
+  let r = prepare_pipe ~close t in
+  Strict_pipe.Reader.fold_without_pushback r
+
+let iter ~close t =
+  let r = prepare_pipe ~close t in
+  Strict_pipe.Reader.iter r
+
+let iter_without_pushback ~close t =
+  let r = prepare_pipe ~close t in
+  Strict_pipe.Reader.iter_without_pushback r
+
+(*
+ * 1. Cached value is keeping peek working
+ * 2. Multiple listeners receive the first mvar value
+ * 3. Multiple listeners receive updates after changes
+ * 4. Listeners can be closed, and other listeners still work
+ * 5. Peek sees the latest value
+ * 6. If we close the shared_mvar, all listeners stop
+*)
+let%test_unit "listeners properly receive updates" =
+  Backtrace.elide := false ;
+  Async.Scheduler.set_record_backtraces true ;
```

Comment:
Should this stuff be committed? I don't know what the overhead is for async backtraces

Reply:
It should be fine for unit tests.